### PR TITLE
feat(firestore): Add new functions and correct old ones

### PR DESCRIPTION
**Pull request by bhshkh**
This PR significantly expands the Firestore Pipelines API by adding a wide range of new expressions and functions. These additions improve parity with other Firestore SDKs and provide developers with more powerful tools for
  complex data transformation and analysis within Firestore.

  Key Changes:

   * New Aggregations: Added RawAggregate to allow calling any aggregation function supported by the Firestore backend, even if not explicitly defined in the SDK. Java [ref](https://github.com/googleapis/java-firestore/blob/6acac4b0598891f66174e911ccf55375b587d773/google-cloud-firestore/src/main/java/com/google/cloud/firestore/pipeline/expressions/AggregateFunction.java#L47-L49).
   * Arithmetic Expressions: Added RoundToPrecision for rounding to specific decimal places. Java [ref](https://github.com/googleapis/java-firestore/blob/6acac4b0598891f66174e911ccf55375b587d773/google-cloud-firestore/src/main/java/com/google/cloud/firestore/pipeline/expressions/Expression.java#L4711-L4761)
   * Array Functions:
       * Offset: Access array elements by index (supports negative indexing).
       * ArrayTransform & ArrayTransformWithIndex: Map array elements to new values using expressions.
       * ArrayFilter: Filter array elements based on a condition (now fully enabled).
   * Reference & Key Functions:
       * GetParent: Retrieve the parent reference of a document or field. 
       * ReferenceSlice & ReferenceSliceToEnd: Extract segments from a document reference path.
   * Logical & General Functions:
       * Coalesce: Return the first non-null value from a list of expressions. Java [ref](https://github.com/googleapis/java-firestore/blob/6acac4b0598891f66174e911ccf55375b587d773/google-cloud-firestore/src/main/java/com/google/cloud/firestore/pipeline/expressions/Expression.java#L318-L348)
       * Variable: Reference variables within lambda-like expressions (e.g., in ArrayTransform).
       * StorageSize: Calculate the storage size of a field or expression.
       * RawBooleanFunction: Support for custom boolean evaluation functions. java [ref](https://github.com/googleapis/java-firestore/blob/6acac4b0598891f66174e911ccf55375b587d773/google-cloud-firestore/src/main/java/com/google/cloud/firestore/pipeline/expressions/BooleanExpression.java#L87-L90)
   * Improved Type Safety: Refined SwitchOn and MapSet signatures to enforce required arguments, reducing runtime errors.
   * Internal Refinements:
       * Added validation helpers for timestamp units, granularity, and parts.
       * Unified internal proto conversion logic for binary and numeric operations.

### CODE REVIEW

File: firestore/pipeline_utils.go [hunk 1]
```diff
@@ -118,6 +111,20 @@ func toExprOrConstant(val any) Expression {
...
 }
 
+// toExprOrNumericConstant converts a plain Go value or an existing Expr into an Expr.
+// Plain values are wrapped in a Constant.
+func toExprOrNumericConstant(val any) Expression {
+	if expr, ok := val.(Expression); ok {
+		return expr
+	}
+	switch v := val.(type) {
+	case int, int8, int16, int32, int64, uint8, uint16, uint32, float32, float64:
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/pipeline_utils.go:121]: @gemini-code-assist 

This is intentional because Firestore uses an int64 to represent integral values, and those types can't be properly represented in an int64.

File: firestore/pipeline_function.go [hunk 2]
```diff
@@ -201,6 +201,18 @@ func Round(numericExprOrFieldPath any) Expression {
 	return newBaseFunction("round", []Expression{asFieldExpr(numericExprOrFieldPath)})
 }
 
+// RoundToPrecision creates an expression that rounds a number to a specified number of decimal places.
+// If places is positive, rounds off digits to the left of the decimal point.
+// If places is negative, rounds off digits to the right of the decimal point.
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/pipeline_function.go]: @gemini-code-assist
Fixed

File: firestore/pipeline_function.go [hunk 3]
```diff
@@ -652,12 +740,16 @@ func CharLength(exprOrFieldPath any) Expression {
...
 //
 // Experimental: Firestore Pipelines is currently in preview and is subject to potential breaking changes in future versions,
 // regardless of any other documented package stability guarantees.
 func StringConcat(exprOrFieldPath any, otherStrings ...any) Expression {
-	return newBaseFunction("string_concat", append([]Expression{asFieldExpr(exprOrFieldPath)}, toExprs(otherStrings)...))
+	exprs := make([]Expression, len(otherStrings))
+	for i, v := range otherStrings {
+		exprs[i] = asStringExpr(v)
+	}
+	return newBaseFunction("string_concat", append([]Expression{asFieldExpr(exprOrFieldPath)}, exprs...))
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/pipeline_function.go]: @gemini-code-assist done

File: firestore/pipeline_function.go [hunk 4]
```diff
@@ -1005,34 +1118,52 @@ func IfErrorBoolean(tryExpr BooleanExpression, catchExpr BooleanExpression) Bool
...
+// - others are optional additional expressions or values to check.
 //
 // Experimental: Firestore Pipelines is currently in preview and is subject to potential breaking changes in future versions,
 // regardless of any other documented package stability guarantees.
-func IfNull(exprOrValue any, others ...any) Expression {
-	return newBaseFunction("if_null", append([]Expression{toExprOrConstant(exprOrValue)}, toArrayOfExprOrConstant(others)...))
+func Coalesce(exprOrField any, replacement any, others ...any) Expression {
+	exprs := make([]Expression, 0, len(others)+2)
+	exprs = append(exprs, asFieldExpr(exprOrField), toExprOrConstant(replacement))
+	exprs = append(exprs, toArrayOfExprOrConstant(others)...)
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/pipeline_function.go]: @gemini-code-assist done

File: firestore/pipeline_function.go [hunk 5]
```diff
@@ -1044,8 +1175,11 @@ func IfNull(exprOrValue any, others ...any) Expression {
 //
 // Experimental: Firestore Pipelines is currently in preview and is subject to potential breaking changes in future versions,
 // regardless of any other documented package stability guarantees.
-func SwitchOn(args ...any) Expression {
-	return newBaseFunction("switch_on", toArrayOfExprOrConstant(args))
+func SwitchOn(condition BooleanExpression, result any, others ...any) Expression {
+	exprs := make([]Expression, 0, len(others)+2)
+	exprs = append(exprs, condition, toExprOrConstant(result))
+	exprs = append(exprs, toArrayOfExprOrConstant(others)...)
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/pipeline_function.go]: @gemini-code-assist done

File: firestore/pipeline_function.go [hunk 6]
```diff
@@ -1095,12 +1229,17 @@ func MapRemove(exprOrField any, strOrExprkey any) Expression {
...
+// - moreKeysAndValues: Optional additional alternating key and value arguments.
 //
 // Experimental: Firestore Pipelines is currently in preview and is subject to potential breaking changes in future versions,
 // regardless of any other documented package stability guarantees.
-func MapSet(exprOrField any, keysAndValues ...any) Expression {
-	return newBaseFunction("map_set", append([]Expression{asFieldExpr(exprOrField)}, toArrayOfExprOrConstant(keysAndValues)...))
+func MapSet(exprOrField any, key any, value any, moreKeysAndValues ...any) Expression {
+	exprs := make([]Expression, 0, len(moreKeysAndValues)+3)
+	exprs = append(exprs, asFieldExpr(exprOrField), toExprOrConstant(key), toExprOrConstant(value))
+	exprs = append(exprs, toArrayOfExprOrConstant(moreKeysAndValues)...)
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply firestore/pipeline_function.go]: @gemini-code-assist done

File: firestore/pipeline_aggregate.go [hunk 7]
```diff
@@ -102,6 +106,21 @@ type AliasedAggregate struct {
...
+//
+// This method provides a way to call aggregation functions that are supported by the Firestore
+// backend but that are not available as specific factory methods in this class.
+//
+// Example:
+//
+//	RawAggregate("sum", "orderAmount").As("totalRevenue")
+//
+// Experimental: Firestore Pipelines is currently in preview and is subject to potential breaking changes in future versions,
+// regardless of any other documented package stability guarantees.
```
Comment (daniel-sanche) [inline firestore/pipeline_aggregate.go:119]: I don't think this is true for pipelines in general anymore, just specific stages/expressions

File: firestore/pipeline_expression.go [hunk 8]
```diff
@@ -663,6 +724,7 @@ func (b *baseExpression) Concat(others ...any) Expression { return Concat(b, oth
 // Key functions
 func (b *baseExpression) GetCollectionID() Expression { return GetCollectionID(b) }
 func (b *baseExpression) GetDocumentID() Expression   { return GetDocumentID(b) }
+func (b *baseExpression) GetParent() Expression       { return GetParent(b) }
```
Comment (daniel-sanche) [inline firestore/pipeline_expression.go]: nit: It might be better to put GetParent and ReferenceSlice into a `// Reference functions` block. That's how [the docs organize it](https://docs.cloud.google.com/firestore/native/docs/pipeline/functions/reference_functions)

File: firestore/pipeline_integration_test.go [hunk 9]
```diff
@@ -2583,6 +2662,22 @@ func keyFuncs(t *testing.T) {
...
+		},
+		{
+			name:     "ReferenceSlice (to end)",
+			pipeline: client.Pipeline().Collection(coll.ID).Select(Fields(ReferenceSlice(ConstantOf(docRef1), 0, 2).As("segments"))),
+			want:     map[string]interface{}{"segments": docRef1.Path},
+		},
+		{
+			name:     "ReferenceSlice (subdoc to doc)",
+			pipeline: client.Pipeline().Collection(coll.ID).Select(Fields(ReferenceSlice(ConstantOf(subDocRef1), 0, 2).As("segments"))),
+			// Note: reference_slice appears to return the full path regardless of length in this test.
```
Comment (daniel-sanche) [inline firestore/pipeline_integration_test.go]: Is this intentional? Or is this a gemini artiffact :)

File: firestore/pipeline_integration_test.go [hunk 10]
```diff
@@ -2583,6 +2662,22 @@ func keyFuncs(t *testing.T) {
 			pipeline: client.Pipeline().Collection(coll.ID).Select(Fields(GetDocumentID(docRef1).As("documentId"))),
 			want:     map[string]interface{}{"documentId": "doc1"},
 		},
+		{
+			name:     "GetParent",
+			pipeline: client.Pipeline().Collection(coll.ID).Select(Fields(GetParent(subDocRef1).As("parent"))),
+			want:     map[string]interface{}{"parent": docRef1.Path},
+		},
```
Comment (daniel-sanche) [inline firestore/pipeline_integration_test.go:2724]: I'm not seeing any tests that return the parent of an un-nested document (should reference the database itself) or the parent of the database (should be null). That would be good to cover. See the [examples here](https://docs.cloud.google.com/firestore/native/docs/pipeline/functions/reference_functions#reference_slice)

I mention this out because it ended up being a major complication in Python. Our parser has classes to handle document references and collection references, but nothing (besides the client) to reference the database itself, so it would crash. Just wanted to point that out in case you run into the same issue

File: firestore/pipeline_integration_test.go [hunk 11]
```diff
@@ -2583,6 +2662,22 @@ func keyFuncs(t *testing.T) {
...
+		{
+			name:     "ReferenceSlice (to end)",
+			pipeline: client.Pipeline().Collection(coll.ID).Select(Fields(ReferenceSlice(ConstantOf(docRef1), 0, 2).As("segments"))),
+			want:     map[string]interface{}{"segments": docRef1.Path},
+		},
```
Comment (daniel-sanche) [inline firestore/pipeline_integration_test.go:2749]: Like my comment for parent, you might want to add more edge case tests, since those edge cases caused issues in python

File: firestore/pipeline_aggregate.go (hunk 7 above)
Comment (bhshkh) [reply firestore/pipeline_aggregate.go:119]: Will remove in next PR

File: firestore/pipeline_integration_test.go (hunk 11 above)
Comment (bhshkh) [reply firestore/pipeline_integration_test.go:2749]: added

File: firestore/pipeline_integration_test.go (hunk 10 above)
Comment (bhshkh) [reply firestore/pipeline_integration_test.go:2724]: added tests and cleanly throw errror

File: firestore/pipeline_integration_test.go (hunk 9 above)
Comment (bhshkh) [reply firestore/pipeline_integration_test.go]: Removed

File: firestore/pipeline_expression.go (hunk 8 above)
Comment (bhshkh) [reply firestore/pipeline_expression.go]: Moved

**Review by daniel-sanche — APPROVED** [review]
LGTM
